### ISSUE TITLE
How persistent is hideRow supposed to be?

### ISSUE THREAD

**Issue by LineWolff** [wenzhixin/bootstrap-table#1715]
It seems like rows hidden with the hideRow method is not persistent through searches using bootstrap-table's own search field. I've made a fiddle to demonstrate.

http://jsfiddle.net/3qLn65vt/

Can anyone shed light on this issue?

**Comment by dabros** [comment]
@LineWolff - not entirely sure the decision behind this myself, but i can see this from the code in current version

```javascript
    // NOTE - called from both showRow and hideRow
    BootstrapTable.prototype.toggleRow = function (index, uniqueId, visible) {
        if (index === -1) {
            return;
        }

        this.$body.find(typeof index !== 'undefined' ?
            sprintf('tr[data-index="%s"]', index) :
            sprintf('tr[data-uniqueid="%s"]', uniqueId))
            [visible ? 'show' : 'hide']();
    };
```

That basically just calls $.show() or $.hide() on that row. It doesnt store any referance to this, and following `onSearch` sees it filter on the `data` array for that table, then call `initBody` which inserts formatted rows for each data item it is passed.

To maintain visible state, youd only need to check it there, at start of data loop in `initBody`.

There does seem to be a persistence in 'visibleColumns', with code related to that in several locations.

If you feel that row visibility should be persistant (if called explicit, optionally until explicit refresh) then a **PR** is always welcome, and ample code for how columns are maintained already exists.

https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#pull-requests

It might also help if you could provide more context as to why this needs to be persistent in your use case, and ideally other use cases you can think of. Be convincing enough and you might get help in implementing this.

https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#feature-requests

**Comment by LineWolff** [comment]
TL;DR: For client-side-only manipulation of visible rows, use filterBy (even though you may need to generate some hidden filter-columns in your data array) instead of hideRow/showRow.

*Thanks* for the explanation, which sheds light on the behavior and led me to realize I had misused the hideRow/showRow method.

I did indeed call the method explicitly in my own filtering/comparison function, and therefore expected a persistent behavior.

I should have used the filterBy method instead, which in my case solved the problem completely. I just hadn't realized I could use that method for the filtering I needed doing. 

Only drawback with filterBy is I need to traverse my data set before loading, adding some hidden columns I can filter off of, but it's a fairly quick and easy operation and really not a problem. On the other hand I benefit from much better performance, since I don't need to reload data into the table after every filter event. And I get the benefits from automatic updating of pagination etc (which I also struggled with, using hideRow/showRow).
